### ISSUE TITLE
v.4.x AssertionFailedException in Decompiler.IL.ControlFlow.LoopDetection.IncludeUnreachablePredecessors

### ISSUE THREAD

**Issue by greenozon** [icsharpcode/ILSpy#1353]
ILSpy version 4.0.0.4468-beta3

```
Error decompiling ---.ScanLiteralString
 ---> ICSharpCode.ILSpy.AssertionFailedException: All jumps into the loop body should go through the entry point
   at ICSharpCode.ILSpy.ILSpyTraceListener.Fail(String message, String detailMessage) in offset 380
   at ICSharpCode.ILSpy.ILSpyTraceListener.Fail(String message) in offset 9
   at System.Diagnostics.TraceInternal.Fail(String message) in offset 55
   at ICSharpCode.Decompiler.IL.ControlFlow.LoopDetection.IncludeUnreachablePredecessors(List`1 loop) in offset 94
   at ICSharpCode.Decompiler.IL.ControlFlow.LoopDetection.Run(Block block, BlockTransformContext context) in offset 425
   at ICSharpCode.Decompiler.IL.Block.RunTransforms(IEnumerable`1 transforms, BlockTransformContext context) in offset 69
   at ICSharpCode.Decompiler.IL.Transforms.BlockILTransform.VisitBlock(ControlFlowNode cfgNode, BlockTransformContext context) in offset 146
   at ICSharpCode.Decompiler.IL.Transforms.BlockILTransform.VisitBlock(ControlFlowNode cfgNode, BlockTransformContext context) in offset 89
   at ICSharpCode.Decompiler.IL.Transforms.BlockILTransform.VisitBlock(ControlFlowNode cfgNode, BlockTransformContext context) in offset 89
   at ICSharpCode.Decompiler.IL.Transforms.BlockILTransform.VisitBlock(ControlFlowNode cfgNode, BlockTransformContext context) in offset 89
   at ICSharpCode.Decompiler.IL.Transforms.BlockILTransform.VisitBlock(ControlFlowNode cfgNode, BlockTransformContext context) in offset 89
   at ICSharpCode.Decompiler.IL.Transforms.BlockILTransform.Run(ILFunction function, ILTransformContext context) in offset 146
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DecompileBody(IMethod method, EntityDeclaration entityDecl, DecompileRun decompileRun, ITypeResolveContext decompilationContext) in offset 581
-- continuing with outer exception (ICSharpCode.Decompiler.DecompilerException) --
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DecompileBody(IMethod method, EntityDeclaration entityDecl, DecompileRun decompileRun, ITypeResolveContext decompilationContext) in offset 1068
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DoDecompile(IMethod method, DecompileRun decompileRun, ITypeResolveContext decompilationContext) in offset 249
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DoDecompile(ITypeDefinition typeDef, DecompileRun decompileRun, ITypeResolveContext decompilationContext) in offset 747
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DoDecompileTypes(IEnumerable`1 types, DecompileRun decompileRun, ITypeResolveContext decompilationContext, SyntaxTree syntaxTree) in offset 219
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DecompileTypes(IEnumerable`1 types) in offset 201
   at ICSharpCode.Decompiler.CSharp.WholeProjectDecompiler.<>c__DisplayClass30_0.<WriteCodeFilesInProject>b__2(IGrouping`2 file) in offset 68
   at System.Threading.Tasks.Parallel.<>c__DisplayClass17_0`1.<ForWorker>b__1() in offset 589
   at System.Threading.Tasks.Task.InnerInvokeWithArg(Task childTask) in offset 6
   at System.Threading.Tasks.Task.<>c__DisplayClass176_0.<ExecuteSelfReplicating>b__0(Object) in offset 134
```

target:  https://www.sendspace.com/file/lyrdcx
note: this material might be obfuscated, so freel free to close it in case of ...

**Comment by dgrunwald** [comment]
This is another instance of #1242.
IncludeUnreachablePredecessors was supposed to be a workaround for problems due to unreachable code, but it can only work in some limited cases.

**Comment by dgrunwald** [comment]
Fixed in 8117dfff4a2b238a0e2794416c78b9d3656dbfb8.
